src/mrinufft/operators/off_resonance.py

```diff
@@ -0,0 +1,330 @@
+    C : np.ndarray
+        Off-resonance phase map at each time segment center of shape
+        ``(n_time_segments, *fieldmap.shape)``.
+        Array module is the same as input fieldmap.
+    """
+    # default
+    if isinstance(n_bins, (list, tuple)) is False:
+        n_bins = (n_bins, 10)
+
+    # transform to list
+    n_bins = list(n_bins)
+
+    # get backend and device
+    xp = get_array_module(fieldmap)
+
+    # enforce complex field
+    fieldmap = xp.asarray(fieldmap, dtype=xp.complex64)
+
+    # cast arrays to fieldmap backend
+    if xp.__name__ == "torch":
+        is_torch = True
+    else:
+        is_torch = False
+
+    if is_cuda_array(fieldmap):
+        assert CUPY_AVAILABLE, "GPU computation requires Cupy!"
+        xp = cp
+        fieldmap = _to_cupy(fieldmap)
+        readout_time = _to_cupy(readout_time)
+        mask = _to_cupy(mask)
+    else:
+        xp = np
+        fieldmap = _to_numpy(fieldmap)
+        readout_time = _to_numpy(readout_time)
+        mask = _to_numpy(mask)
+
+    readout_time = xp.asarray(readout_time, dtype=xp.float32).ravel()
+    if mask is None:
+        mask = xp.ones_like(fieldmap, dtype=bool)
+    else:
+        mask = xp.asarray(mask, dtype=bool)
+
+    # get field map
+    if xp.isreal(fieldmap).all().item():
+        r2star = None
+        b0 = fieldmap
+        fieldmap = 0.0 + 1j * b0
+    else:
+        r2star = fieldmap.real
+        b0 = fieldmap.imag
+
+    # Hz to radians / s
+    fieldmap = 2 * math.pi * fieldmap
+
+    # create histograms
+    if r2star is not None:
+        z = fieldmap[mask].ravel()
+        z = xp.stack((z.imag, z.real), axis=1)
+        hk, ze = xp.histogramdd(z, bins=n_bins)
+        ze = list(ze)
+
+        # get bin centers
+        zc = [e[1:] - (e[1] - e[0]) / 2 for e in ze]
+
+        # complexify
+        zk = _outer_sum(1j * zc[0], zc[1])  # [K1 K2]
+        zk = zk.T
+        hk = hk.T
+    else:
+        z = fieldmap[mask].ravel()
+        hk, ze = xp.histogram(z.imag, bins=n_bins[0])
+
+        # get bin centers
+        zc = ze[1:] - (ze[1] - ze[0]) / 2
+
+        # complexify
+        zk = 0 + 1j * zc  # [K 1]
+
+    # flatten histogram values and centers
+    hk = hk.ravel()
+    zk = zk.ravel()
+
+    # generate time for each segment
+    tl = xp.linspace(
+        readout_time.min(), readout_time.max(), n_time_segments, dtype=xp.float32
+    )  # time seg centers in [s]
+
+    # prepare for basis calculation
+    ch = xp.exp(-tl[:, None, ...] @ zk[None, ...])
+    w = xp.diag(hk**0.5)
+    p = xp.linalg.pinv(w @ ch.T) @ w
+
+    # actual temporal basis calculation
+    B = p @ xp.exp(-zk[:, None, ...] * readout_time[None, ...])
+    B = B.astype(xp.complex64)
+
+    # get spatial coeffs
+    C = xp.exp(-tl * fieldmap[..., None])
+    C = C[None, ...].swapaxes(0, -1)[
+        ..., 0
+    ]  # (..., n_time_segments) -> (n_time_segments, ...)
+
+    # clean-up of spatial coeffs
+    C = xp.nan_to_num(C, nan=0.0, posinf=0.0, neginf=0.0)
```

Comment:
Should `C` be also casted to np.complex64, as for `B` ?

Reply:
I agree!